sdk/core/azure-core/azure/core/rest/_http_response_impl.py

```diff
@@ -0,0 +1,320 @@
+    responses. This is the base implementation class shared by HttpResponseImpl and AsyncHttpResponseImpl.
+    The transport responses will be built on top of HttpResponseImpl and AsyncHttpResponseImpl
+
+    :keyword request: The request that led to the response
+    :type request: ~azure.core.rest.HttpRequest
+    :keyword any internal_response: The response we get directly from the transport. For example, for our requests
+     transport, this will be a requests.Response.
+    :keyword optional[int] block_size: The block size we are using in our transport
+    :keyword int status_code: The status code of the response
+    :keyword str reason: The HTTP reason
+    :keyword str content_type: The content type of the response
+    :keyword MutableMapping[str, str] headers: The response headers
+    :keyword Callable stream_download_generator: The stream download generator that we use to stream the response.
+    """
+
+    def __init__(self, **kwargs):
+        # type: (Any) -> None
+        super(_HttpResponseBaseImpl, self).__init__()
+        self._request = kwargs.pop("request")
+        self._internal_response = kwargs.pop("internal_response")
+        self._block_size = kwargs.pop("block_size", None) or 4096  # type: int
+        self._status_code = kwargs.pop("status_code")  # type: int
+        self._reason = kwargs.pop("reason")  # type: str
+        self._content_type = kwargs.pop("content_type")  # type: str
+        self._headers = kwargs.pop("headers")  # type: MutableMapping[str, str]
+        self._stream_download_generator = kwargs.pop("stream_download_generator")  # type: Callable
+        self._is_closed = False
+        self._is_stream_consumed = False
+        self._json = None  # this is filled in ContentDecodePolicy, when we deserialize
+        self._content = None  # type: Optional[bytes]
+        self._text = None  # type: Optional[str]
+
+    @property
+    def request(self):
+        # type: (...) -> _HttpRequest
+        """The request that resulted in this response.
+
+        :rtype: ~azure.core.rest.HttpRequest
+        """
+        return self._request
+
+    @property
+    def url(self):
+        # type: (...) -> str
+        """The URL that resulted in this response.
+
+        :rtype: str
+        """
+        return self.request.url
+
+    @property
+    def is_closed(self):
+        # type: (...) -> bool
+        """Whether the network connection has been closed yet.
+
+        :rtype: bool
+        """
+        return self._is_closed
+
+    @property
+    def is_stream_consumed(self):
+        # type: (...) -> bool
+        """Whether the stream has been fully consumed"""
+        return self._is_stream_consumed
+
+    @property
+    def status_code(self):
+        # type: (...) -> int
+        """The status code of this response.
+
+        :rtype: int
+        """
+        return self._status_code
+
+    @property
+    def headers(self):
```

Comment:
For inputs, are we requiring the caller to correctly serialize header values (e.g. date times, numbers etc.)?

Reply:
sorry a little confused by inputs by the caller, since this is not settable

Reply:
It's a mutable mapping, so the caller in this case would be whoever wants to add stuff to the headers collection. Sorry for not being clear. And this goes for headers in general - do we ever accept a value that is not a string (e.g. when making a request)?

Reply:
Ah i see. We do, for headers in azure.core.rest (request or responses), we don't do anything here to make sure that they are strings. If we want to, I can do a separate PR to add that